tests/integration/test_charm.py

```diff
@@ -37,24 +42,248 @@ async def test_build_and_deploy(ops_test: OpsTest):
+
+    # Add a unit and wait until it is active
+    await ops_test.model.applications[APP_NAME].add_unit()
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 4
+
+    added_unit = [
+        unit
+        for unit in ops_test.model.applications[APP_NAME].units
+        if unit.name not in old_unit_names
+    ][0]
+
+    # Ensure that all units have the above inserted data
+    for unit in ops_test.model.applications[APP_NAME].units:
+        output = await run_command_on_unit(
+            unit,
+            f"{mysqlsh_sql_base_command} -e \"SELECT * FROM test.instance_state_replication WHERE id = '{random_chars}'\"",
+        )
+        assert random_chars in output
+
+    # Destroy the recently created unit and wait until the application is active
+    await ops_test.model.destroy_units(added_unit.name)
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+    # Ensure that the data still exists in all the units
+    for unit in ops_test.model.applications[APP_NAME].units:
+        output = await run_command_on_unit(
+            unit,
+            f"{mysqlsh_sql_base_command} -e \"SELECT * FROM test.instance_state_replication WHERE id = '{random_chars}'\"",
+        )
+        assert random_chars in output
+
+    await ops_test.model.set_config({"update-status-hook-interval": "60m"})
+
+
+async def get_primary_unit(
+    ops_test: OpsTest, unit: Unit, server_config_username: str, server_config_password: str
+) -> str:
+    """Helper to retrieve the primary unit.
+
+    Args:
+        ops_test: The ops test object passed into every test case
+        unit: A unit on which to run dba.get_cluster().status() on
+        server_config_username: The server config username
+        server_config_password: The server config password
+
+    Returns:
+        A juju unit that is a MySQL primary
+    """
+    commands = [
+        "mysqlsh",
+        "--python",
+        f"{server_config_username}:{server_config_password}@127.0.0.1",
+        "-e",
+        f"\"print('<CLUSTER_STATUS>' + dba.get_cluster('{CLUSTER_NAME}').status().__repr__() + '</CLUSTER_STATUS>')\"",
+    ]
+    raw_output = await run_command_on_unit(unit, " ".join(commands))
+    if not raw_output:
+        return None
+
+    matches = re.search("<CLUSTER_STATUS>(.+)</CLUSTER_STATUS>", raw_output)
+    if not matches:
+        return None
+
+    cluster_status = json.loads(matches.group(1).strip())
+
+    primary_name = [
+        label
+        for label, member in cluster_status["defaultReplicaSet"]["topology"].items()
+        if member["mode"] == "R/W"
+    ][0].replace("-", "/")
```

Comment:
Might be more readable as a bog-standard `for` loop, `black` formatting for list-comprehension is a pain.

Reply:
Refactored to be a one liner in ff877d16df482a9781067a33b0a1cf1e46591782